Add classic mod for the osu! ruleset

smoogipoo opened the pull request:
Prereqs:
- [x] https://github.com/ppy/osu/pull/11679

Closes https://github.com/ppy/osu/issues/7521.

Tested with both ThePooN and FGSky's replays on https://osu.ppy.sh/beatmapsets/743239#osu/1567435. Both combo break on master and don't with this PR.
To test, you can just `yield return new OsuModClassic()` from `OsuRuleset.ConvertFromLegacyMods()` and re-import the score.

Unsure how we want to apply this just yet, so I've only added it as a mod without forcing it upon replay/beatmaps.

smoogipoo committed:
Add flag to toggle follow circle tracking for slider heads

smoogipoo committed:
Add flags to return to classic slider scoring

smoogipoo committed:
Make sliders display judgements when not ignored

smoogipoo committed:
Add OsuModClassic

smoogipoo committed:
Add object-ordered hit policy

smoogipoo committed:
Add OOHP tests

smoogipoo committed:
Merge branch 'hit-policy-refactor' into classic-mode-flags

smoogipoo committed:
Disable snaking out when tracking is disabled

smoogipoo committed:
Add classic slider ball tracking

smoogipoo committed:
Move follow circle tracking to DrawableSliderHead

smoogipoo commented on osu.Game.Rulesets.Osu.Tests/TestSceneObjectOrderedHitPolicy.cs:
```diff
@@ -0,0 +1,491 @@
+
+namespace osu.Game.Rulesets.Osu.Tests
+{
+    public class TestSceneObjectOrderedHitPolicy : RateAdjustedBeatmapTestScene
```
I can extract this to a separate PR if you'd like - the implementation of `ObjectOrderedHitPolicy`. This file is basically a copy-paste of `TestSceneStartTimeOrderedHitPolicy` with judgements changed and a new method `TestClickSecondCircleAfterFirstCircleTimeWithFirstCircleJudged`.

smoogipoo commented:
In this PR, I've made all booleans have true values by default, with the classic mod then setting them all to `false`. I'm interested in thoughts and suggestions here if this seems off, and especially with how the mod is doing `!DisableX.Value`.

bdach commented on osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSliderHead.cs:
```diff
@@ -59,12 +70,28 @@ protected override void Update()
-                Position = Slider.CurvePositionAt(completionProgress);
+            // If not judged as a normal hitcircle, only track whether a hit has occurred (via IgnoreHit) rather than a scorable hit result.
+            var result = base.ResultFor(timeOffset);
+            return result.IsHit() ? HitResult.IgnoreHit : result;
```
Shouldn't this be

```suggestion
            return result.IsHit() ? HitResult.IgnoreHit : HitResult.IgnoreMiss;
```

bdach commented on osu.Game.Rulesets.Osu/Mods/OsuModClassic.cs:
```diff
@@ -0,0 +1,84 @@
+
+namespace osu.Game.Rulesets.Osu.Mods
+{
+    public class OsuModClassic : Mod, IApplicableToHitObject, IApplicableToDrawableHitObjects, IApplicableToDrawableRuleset<OsuHitObject>
```
This mod doesn't override `Mod.Type` to `Conversion`. Looks a little wrong in the mod list due to that (wrong icon colour).

smoogipoo committed:
Override mod type

smoogipoo committed:
Prevent invalid hit results for ignored slider heads

smoogipoo committed:
Merge branch 'hit-policy-refactor' into classic-mode-flags

peppy committed:
Merge branch 'master' into classic-mode-flags

peppy commented on osu.Game.Rulesets.Osu/Objects/Slider.cs:
```diff
@@ -114,8 +114,14 @@ public int RepeatCount
+        /// Whether this <see cref="Slider"/>'s judgement should be ignored.
+        /// If <c>false</c>, this <see cref="Slider"/> will be judged proportionally to the number of ticks hit.
+        /// </summary>
+        public bool IgnoreJudgement = true;
```
shouldn't this be defaulting to false?

peppy replied:
i.. .guess ... not...

peppy commented on osu.Game.Rulesets.Osu/Skinning/Default/SliderBall.cs:
```diff
@@ -31,6 +31,12 @@ public Color4 AccentColour
+        /// Whether to track accurately to the visual size of this <see cref="SliderBall"/>.
+        /// If <c>false</c>, tracking will be performed at the final scale at all times.
+        /// </summary>
+        public bool TrackVisualSize = true;
```
```suggestion
        public bool InputTracksVisualSize = true;
```

Maybe reads better?

peppy commented on osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSlider.cs:
```diff
@@ -249,7 +250,28 @@ protected override void CheckForResult(bool userTriggered, double timeOffset)
                 return;
 
-            ApplyResult(r => r.Type = NestedHitObjects.Any(h => h.Result.IsHit) ? r.Judgement.MaxResult : r.Judgement.MinResult);
+            if (HitObject.IgnoreJudgement)
```
Hmm, going to have to have a think about this one. I'm not sure about polluting the `DrawableSlider` class with this kind of stable knowledge... Wonder how well this would work as a function we can assign from the mod itself?

If this is to remain here, I'd amend the comment just below to mention that the non-"ignore" case is the legacy implementation, because the variable naming feels a bit inverse otherwise (see my mistaken comment from before this review).

smoogipoo replied:
I've kept it as is and adjusted the comments. I'm not sure how far to go in that case because `DrawableSliderHead` has a similar property added for follow circle tracking.

peppy commented on osu.Game.Rulesets.Osu/Skinning/Default/PlaySliderBody.cs:
```diff
@@ -35,11 +37,29 @@ private void load(ISkinSource skin, DrawableHitObject drawableObject)
+            SnakingOut.BindTo(snakingOut);
             config?.BindWith(OsuRulesetSetting.SnakingInSliders, SnakingIn);
-            config?.BindWith(OsuRulesetSetting.SnakingOutSliders, SnakingOut);
+            config?.BindWith(OsuRulesetSetting.SnakingOutSliders, snakingOut);
```
is this intentionally no longer binding to the bindable of the underlying class? if so it needs explanation (and probably a variable rename of this private instance to better describe why it is different).

peppy replied:
oh, it is just above.. but also creating a private bindable for unknown reasons.

smoogipoo replied:
It's creating a private bindable so that the `TrackFollowCircle` property doesn't affect the configuration bindable. So the config bindable is bound privately, and `SnakingOut` is unbound + mutated locally in the `!TrackFollowCircle` case.

peppy replied:
Right. I guess renaming the private to `configSnakingOut` would help.

peppy committed:
Remove unnecessary initial call to HitObjectApplied bound method

Was causing test failures. Looks to be unnecessary on a check of when
HitObjectApplied is invoked.

peppy commented on osu.Game.Rulesets.Osu/Mods/OsuModClassic.cs:
```diff
@@ -0,0 +1,86 @@
+
+        public override ModType Type => ModType.Conversion;
+
+        [SettingSource("Disable slider head judgement", "Scores sliders proportionally to the number of ticks hit.")]
```
```suggestion
        [SettingSource("No slider head accuracy requirement", "Scores sliders proportionally to the number of ticks hit.")]
```

peppy commented on osu.Game.Rulesets.Osu/Mods/OsuModClassic.cs:
```diff
@@ -0,0 +1,86 @@
+        [SettingSource("Disable slider head judgement", "Scores sliders proportionally to the number of ticks hit.")]
+        public Bindable<bool> DisableSliderHeadJudgement { get; } = new BindableBool(true);
+
+        [SettingSource("Disable slider head tracking", "Pins slider heads at their starting position, regardless of time.")]
```
```suggestion
        [SettingSource("No slider head movement", "Keep slider heads stationary at their starting position, instead of following the ball.")]
```

peppy commented on osu.Game.Rulesets.Osu/Mods/OsuModClassic.cs:
```diff
@@ -0,0 +1,86 @@
+        [SettingSource("Disable slider head tracking", "Pins slider heads at their starting position, regardless of time.")]
+        public Bindable<bool> DisableSliderHeadTracking { get; } = new BindableBool(true);
+
+        [SettingSource("Disable note lock lenience", "Applies note lock to the full hit window.")]
```
```suggestion
        [SettingSource("Apply classic note lock", "Only the very next object in time will be hittable.")]
```

peppy commented on osu.Game.Rulesets.Osu/Mods/OsuModClassic.cs:
```diff
@@ -0,0 +1,86 @@
+        [SettingSource("Disable note lock lenience", "Applies note lock to the full hit window.")]
+        public Bindable<bool> DisableLenientNoteLock { get; } = new BindableBool(true);
+
+        [SettingSource("Disable exact slider follow circle tracking", "Makes the slider follow circle track its final size at all times.")]
```
```suggestion
        [SettingSource("Use fixed slider follow circle hit area", "The valid follow radius is fixed in size, rather than adjusting with the visual scaling animation.")]
```

peppy reviewed (changes requested):
Attempted to make the user-visible explanations read a bit better... hopefully.

Joehuu commented:
Seems to close https://github.com/ppy/osu/issues/7521?

AtakanErmis commented:
Sliderbreaks doesn't work same as osu! stable. It gives 100 if player misses the head but not breaking the combo.

Comparison with stable: [stable](https://youtu.be/avobj1MxTjY) vs [lazer with classic mod](https://www.youtube.com/watch?v=Esdb-uyjqdY)

Map and replay I used for test:

Map: https://osu.ppy.sh/beatmapsets/414034#osu/1171254
Replay file: https://drive.google.com/file/d/1pfyFSK2BsyqtutS4gu78sueA5GSel2x6/view?usp=sharing

smoogipoo commented:
@AtakanErmis Thanks for that. I'd completely forgotten about combo and tunnel-visioned onto scoring... It needs a little more thought since we don't currently have a way to affect combo but not score. The closest we can get right now is awarding 30 score via `LargeTick` judgements but that's still different to stable.

@peppy what's your thoughts on this? In short, we've got `SmallTick` that affects only the accuracy portion, and `SmallBonus`/`LargeBonus` that affect only the bonus portion, but nothing that affects only the combo portion: https://github.com/ppy/osu/wiki/Scoring#score .
We can either go down the path of using `LargeTick` and awarding 30 score for hits within any of the non-miss window, and accept that there's a difference in scoring, or add a `ComboHit`/`ComboMiss` hit result specifically for this case.

smoogipoo commented:
Actually nevermind, because stable already adds 30 score for correctly-hit heads, so `LargeTick` is the proper way to go.

smoogipoo commented:
@AtakanErmis I can't access that gdrive file, you'll need to enable public access.

smoogipoo committed:
Judge heads as slider ticks instead

smoogipoo committed:
Apply review comments (user explanations + property names)

smoogipoo committed:
Rename bindable member, reorder binds

smoogipoo committed:
Adjust comment

smoogipoo committed:
Rename IgnoreJudgement -> OnlyJudgeNestedObjects

smoogipoo committed:
Adjust DrawableSlider comment

smoogipoo committed:
Rename SliderBall flag

smoogipoo committed:
Adjust comment for DrawableSliderHead

smoogipoo commented:
Thanks :) Looks good now.

smoogipoo committed:
Fix unintended changes due to renaming or otherwise

smoogipoo committed:
Update test

adpii commented:
Sliders give 100 even if a player completely misses the slider when "No slider head accuracy requirement" is enabled

https://user-images.githubusercontent.com/17258921/107507880-440cb900-6bb1-11eb-9935-f597faabe582.mp4

Map: https://osu.ppy.sh/beatmapsets/414034#osu/1171254

smoogipoo committed:
Fix div-by-0 when 0 ticks are hit

smoogipoo committed:
Simplify case

smoogipoo committed:
Fix grammatical error in comment

adpii commented on osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSlider.cs:
```diff
@@ -249,7 +250,30 @@ protected override void CheckForResult(bool userTriggered, double timeOffset)
+                    r.Type = HitResult.Miss;
+                else
+                {
+                    double hitFraction = (double)totalTicks / hitTicks;
```
I was also wondering why sliders don't ever give 50s, and this is why

```suggestion
                    double hitFraction = (double)hitTicks / totalTicks;
```

smoogipoo replied:
Jeez... What was I thinking. Good catch.

This could use a test, but I'm not sure whether that's more work than it's worth at this point.

smoogipoo committed:
Fix inverted calculation

peppy committed:
Merge branch 'master' into classic-mode-flags

peppy commented:
Will need support added at osu-web end.
